### PR TITLE
feat: experimental i18n routing

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

Experimental i18n routing.

RFC: https://github.com/withastro/roadmap/pull/734

## Testing

New tests added. The existing tests not related to i18n routing should pass because this feature is **opt-in**.

## Docs

/cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

**Comment by ematipico** [comment]
!preview i18n-routing

**Comment by ematipico** [comment]
!preview i18n-routing

File: packages/astro/src/@types/astro.ts [hunk 1]
```diff
@@ -1375,6 +1375,72 @@ export interface AstroUserConfig {
...
+             * @default {'prefix-other-locales'}
+             * @version 3.*.*
+             * @description
+             *
+             * Controls the routing strategy:
+             *  - `prefix-other-locales`: This is the default value. When used, Astro will identify those routes that belong to the default locale and execute a redirect to a route stripped of that locale.
+             *		Use `example.com/[lang]/content/` for every route.
+             *  - `prefix-always`: When used, all URLs need to contains a locale prefix. Astro will return a 404 for any route that doesn't fullfill the requirements.
+             * 		Use `example.com/[lang]/content/` for all non-default languages, but `example.com/content` for the default locale/.
+             *
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1440]:
```suggestion
             * Controls the routing strategy to determine your site URLs.
             *
             *  - `prefix-other-locales` (default): Only non-default languages will display a language prefix. The `defaultLocale` will not show a language prefix.
             * URLs will be of the form `example.com/[lang]/content/` for all non-default languages, but `example.com/content/` for the default locale.
     
             *  - `prefix-always`: All URLs will display a language prefix.
             *  URLs will be of the form`example.com/[lang]/content/` for every route, including the default language.
             *
             * Note: Astro requires all content to exist within a `/[lang]/` folder, even for the default language. When configured to `prefix-other-locales`, Astro will execute a redirect to a route stripped of that locale.
```
This is significantly different, and I don't even think it's necessarily correct, but this is the kind of thing we have to be able to write, so I'll need you to fix this up in this format.  😅 

It's important here in this config setting that people are choosing based on what URLs they are going to get, because that's what they can see and that's what they know. They are going to choose this strategy not based on things like redirects behind the scenes, but based on the URL patterns they want.

So, I've reconstructed these to show which URL patterns you'll get if you choose each option. I've put the "how it works under the hood" separately below the options (and, I'm not even sure this is totally correct, so please correct!). This might not even necessary information to have here, really, as it can go in longer-form docs where we can explain how they need to structure their content. This is OK even if it just says when you configure X, you get these URLs. When you configure Y, you get these URLs.

**Review by sarah11918 — COMMENTED** [review]
OK, I've left some feedback on the docs and error messages here! If there's something that I should have looked at but didn't, then please let me know!

ALSO: I may come back and update these as I read the docs PR. I happened to do this page first, as it was less and more concise content. I may read/learn something in the larger guide that either clears some things up for me here, or changes my opinion on some things. So, if you're thinking, "Didn't she read the guide??" welll, no, not yet! Gotta start somewhere!

File: packages/astro/src/@types/astro.ts [hunk 2]
```diff
@@ -1375,6 +1375,72 @@ export interface AstroUserConfig {
 		 * ```
 		 */
 		devOverlay?: boolean;
+
+		// TODO review with docs team before merging to `main`
+		/**
+		 * @docs
+		 * @name experimental.i18n
+		 * @type {object}
```
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:1383]: Exactly, no plain boolean. We talked about it internally, and we're happy to ship the configuration like this because it's possible we are going to change the names of the configurations too.

File: packages/astro/src/@types/astro.ts [hunk 3]
```diff
@@ -1375,6 +1375,72 @@ export interface AstroUserConfig {
...
+		 */
+		i18n?: {
+			/**
+			 * @docs
+			 * @name experimental.i18n.defaultLocale
+			 * @type {string}
+			 * @version 3.*.*
+			 * @description
+			 *
+			 * The default locale of your website/application
```
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:1398]: > This should probably also give guidance like what form/system you expect, even with an external link, so that people know what syntax to use here.
> 

We haven't mentioned this in the RFC, it would be great if you could add a comment there. We haven't set any restrictions, what do you suggest?

> I'd also indicate whether this is required.

Do I need to use a special JSDoc tag? Or should we just say it in the description?

File: packages/astro/src/@types/astro.ts [hunk 4]
```diff
@@ -1375,6 +1375,72 @@ export interface AstroUserConfig {
...
+			 * @name experimental.i18n.fallback
+			 * @type {Record<string, string[]>}
+			 * @version 3.*.*
+			 * @description
+			 *
+			 * The fallback system of the locales. By default, the fallback system affect the **content only**, and it doesn't
+			 * do any redirects.
+			 *
+			 * This means that when attempting to navigate to a page that hasn't been translated, Astro will pull the content
+			 * from the page of the default locale and render it. No redirects will happen.
```
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:1423]: The documentation of this part needs to be updated, and it's on me. I am going to update it first, and I will ask you to review it again.

File: packages/astro/src/@types/astro.ts (hunk 1 above)
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:1440]: Like your format, just one thing that changed recently: https://github.com/withastro/roadmap/pull/734/commits/989ecc1c40eb4ad43acea78be3212ea3f688f974

Astro won't do any redirects in `prefix-always`, so URLs like `example.com/content` will result in a 404. I let you rephrase this.

File: packages/astro/src/core/errors/errors-data.ts [hunk 5]
```diff
@@ -1272,5 +1272,21 @@ export const UnsupportedConfigTransformError = {
...
+		return `The locale \`${locale}\` does not exist in the configured locales. Available locales: ${locales.join(
+			', '
+		)}.`;
+	},
+} satisfies ErrorData;
+
+export const CantRenderPage = {
+	name: 'CantRenderPage',
+	title: "Astro can't render the route.",
+	message: "You tried to render a route that isn't a redirect and doesn't have any component.",
```
[the remark this answers is not in the document]
Comment (sarah11918) [reply packages/astro/src/core/errors/errors-data.ts:1288]: Hmm... The hint could maybe be something like:

> If you expect to find a route here, this may be an Astro bug. Please file an issue/restart the dev server/...

Or something like that. It would be helpful to stop a user from checking and double checking all their code if there's nothing that they can actually do here. Because I think the natural thing to do when you get an error is try to change something in your code, so if we can stop them from doing that, that's helpful.

File: packages/astro/src/@types/astro.ts (hunk 1 above)
Comment (sarah11918) [reply packages/astro/src/@types/astro.ts:1440]: Sounds good! I'll edit the above suggestion to include that information, then you can fix whatever needs fixing!

Comment (sarah11918) [reply packages/astro/src/@types/astro.ts:1440]: Suggestion updated! You can now work from that!

File: packages/astro/src/core/errors/errors-data.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarah11918) [reply packages/astro/src/core/errors/errors-data.ts:1288]: As for the message itself, maybe:

> You attempted to render a route that does not exist. There is no file or redirect associated with this route."

File: packages/astro/src/@types/astro.ts [hunk 6]
```diff
@@ -1375,6 +1375,75 @@ export interface AstroUserConfig {
...
+			/**
+			 * @docs
+			 * @name experimental.i18n.defaultLocale
+			 * @type {string}
+			 * @version 3.*.*
+			 * @description
+			 *
+			 * The default locale of your website/application
+			 */
+			defaultLocale: string;
```
Comment (florian-lefebvre) [inline packages/astro/src/@types/astro.ts:1400]: Could we make it a generic so that we can enforce it in `locales` at the type level?

File: packages/astro/src/core/app/index.ts [hunk 7]
```diff
@@ -204,17 +222,27 @@ export class App {
 		page: SinglePageBuiltModule,
 		status = 200
 	): Promise<RenderContext> {
+		let preferredLocale: undefined | string = undefined;
+		let preferredLocaleList: undefined | string[] = undefined;
+		if (this.#manifest.i18n) {
+			const result = computePreferredLocales(request, this.#manifest.i18n.locales);
```
Comment (matthewp) [inline packages/astro/src/core/app/index.ts:228]: As we talked about on the call, it would be good to compute these lazily only when the user asks for it. First, this will improve performance as we won't need to parse this in sites not using it, but also the warning we get in SSG will not occur.

To do this you'll need to pass the `locales` list down to here: https://github.com/withastro/astro/blob/main/packages/astro/src/core/render/result.ts#L187

That line is where cookies are parsed lazily. So something like this will be needed.

File: packages/astro/src/@types/astro.ts [hunk 8]
```diff
@@ -1375,6 +1375,91 @@ export interface AstroUserConfig {
...
+			/**
+			 * @docs
+			 * @name experimental.i18n.locales
+			 * @type {string[]}
+			 * @version 3.5.0
+			 * @description
+			 *
+			 * A list of all locales supported by the website. e.g. `['en', 'fr', 'es']` This list should/should not include the `defaultLocale`.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1408]:
```suggestion
			 * A list of all locales supported by the website. e.g. `['en', 'fr', 'es']` This list should also include the `defaultLocale`.
```
Just calling out that we need to pick one!  😄 

I'm assuming from the bug bash that `defaultLocale` SHOULD be included in this list, so that's the edit I'm suggesting. If not, fix it the other way!

I will also note that once we define a system/syntax, we should mention that same one both here and in `defaultLocale`.

QUESTION: Is it true that right now you don't really *care* what syntax is used, as long as the folder structure matches what's listed here?

E.g. You could make up a language `dqwx` and as long as you have files in `src/pages/dqwx/` the routing system functions properly?

In that case, I'd suggest wording like:

> No particular language format or syntax is enforced, but your folder structure must match exactly the `locales` in the list.

File: packages/astro/src/@types/astro.ts [hunk 9]
```diff
@@ -1375,6 +1375,91 @@ export interface AstroUserConfig {
...
+			/**
+			 * @docs
+			 * @name experimental.i18n.fallback
+			 * @type {Record<string, string>}
+			 * @version 3.5.0
+			 * @description
+			 *
+			 * The content fallback strategy when navigating to pages in non-default languages that do not exist. (e.g. a translated page has not been created).
+			 *
+			 * This a plain object, where a key is a locale that hasn't traslated pages, and the corresponding value is the locale where you want to redirect the missing page to.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1421]:
```suggestion
			 * The fallback strategy when navigating to pages that do not exist (e.g. a translated page has not been created).
			 *
			 * Use this object to declare a fallback `locale` route for each language you support. If no fallback is specified, then unavailable pages will return a 404.
```
Just notice this suggestion didn't get merged in the other PR, and I'm suggesting again another tiny tweak to it:

I don't think introducing the idea of "content" here is actually needed! Now that I've seen how this works, and especially given we're talking about which pages to navigate to or to 404, I don't think this is a "content" strategy. This allows us to save that word in case you do end up doing something like the docs that *does* inject some fallback content on a page (at the requested route).

I also think we can simplify the wording by eliminating "non-default languages" and just say, what to do when a page doesn't exist. Which is a nicer sentence for everyone!

So what do you think about the above?

**Comment by ematipico** [comment]
@sarah11918 
- I added the changelog: https://github.com/withastro/astro/pull/8974/commits/217d844f2fa590b956c73e9be823ba75cba1d956
- I applied your suggestions: https://github.com/withastro/astro/pull/8974/commits/1a467919cacf1aed4d1a3462a89ff8b4ad402464

File: packages/astro/src/@types/astro.ts (hunk 6 above)
Comment (matthewp) [reply packages/astro/src/@types/astro.ts:1400]: We were planning on adding better type support through some type generation.

Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:1400]: yeah codegen is probably easier than generics in this case

File: .changeset/rude-lizards-scream.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+You have access to two new properties in the `Astro` global: `Astro.preferredLocale` and `Astro.preferredLocaleList`. These are locales computed from the `Accept-Langauge` header, and supported by your website. You will never receive locales that aren't configured in your website.
+
+```astro
+---
+const preferredLocale = Astro.preferredLocale;
+const preferredLocaleList = Astro.preferredLocaleList;
+---
+```
```
Comment (sarah11918) [inline .changeset/rude-lizards-scream.md:42]:
``````suggestion
---
'astro': minor
---

Experimental support for i18n routing. 

Astro's experimental i18n routing allows you to add your multilingual content with support for configuring a default language and URL path convention, computing relative page URLs, and accepting preferred languages provided by your visitor's browser browser.

Enable the experimental routing option by adding an `i18n` object to your Astro configuration with a default location and a list of all languages to support:

```js
// astro.config.mjs
import {defineConfig} from "astro/config";

export default defineConfig({
    experimental: {
        i18n: {
            defaultLocale: "en",
            locales: ["en", "es", "pt_BR"]
        }
    }
})
```

Organize your content folders by locale, including your `defaultLocale` and `src/pages/index.astro` will now automatically default to the `index.astro` file of your default language.

```
├── src
│   ├── pages
│   │   ├── en
│   │   │   ├── about.astro
│   │   │   ├── index.astro
│   │   ├── es
│   │   │   ├── about.astro
│   │   │   ├── index.astro
│   │   ├── pt_BR
│   │   │   ├── about.astro
│   │   │   ├── index.astro
│   ├── index.astro

```

Compute relative URLs for your links with `getLocaleRelativeURL` from the new `astro:i18n` module:

```astro
---
import {getLocaleRelativeUrl} from "astro:i18n";
const aboutUrl = getLocaleRelativeUrl("pt_Br", "about");
---
<p>Learn more <a href={aboutURL}>About</a> this site!</p>
```

Enabling i18n routing also provides two new properties for browser language detection: `Astro.preferredLocale` and `Astro.preferredLocaleList`. These combine the browser's `Accept-Langauge` header, and your site's list of supported languages and can be used to automatically respect your visitor's preferred languages.

```astro
---
const preferredLocale = Astro.preferredLocale;
const preferredLocaleList = Astro.preferredLocaleList;
---
NEED A USAGE EXAMPLE HERE
```

Read more about Astro's [experimental i18n routing](https://docs.astro.build/en/guides/i18n-routing/#browser-language-detection) in our documentation.
``````

File: packages/astro/src/@types/astro.ts [hunk 11]
```diff
@@ -1375,6 +1375,93 @@ export interface AstroUserConfig {
...
+			 */
+			defaultLocale: string;
+			/**
+			 * @docs
+			 * @name experimental.i18n.locales
+			 * @type {string[]}
+			 * @version 3.5.0
+			 * @description
+			 *
+			 * A list of all locales supported by the website. e.g. `['en', 'fr', 'es']` This list should also include the `defaultLocale`. This is a required field.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1408]:
```suggestion
			 * A list of all locales supported by the website (e.g. `['en', 'es', 'pt_BR']`). This list should also include the `defaultLocale`. This is a required field.
```
Fix some punctuation and it's probably good to show that they don't HAVE to be two letter codes.

File: packages/astro/src/@types/astro.ts [hunk 12]
```diff
@@ -1375,6 +1375,93 @@ export interface AstroUserConfig {
...
+			 * @name experimental.i18n.fallback
+			 * @type {Record<string, string>}
+			 * @version 3.5.0
+			 * @description
+			 *
+			 * The fallback strategy when navigating to pages that do not exist (e.g. a translated page has not been created).
+			 *
+			 * Use this object to declare a fallback `locale` route for each language you support. If no fallback is specified, then unavailable pages will return a 404.
+			 *
+			 * #### Examples
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1425]:
```suggestion
			 * #### Example
```
nit: There's only one example shown! 😄

File: .changeset/rude-lizards-scream.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+```astro
+---
+const preferredLocale = Astro.preferredLocale;
+const preferredLocaleList = Astro.preferredLocaleList;
+---
+NEED A USAGE EXAMPLE HERE
+```
+
+Read more about Astro's [experimental i18n routing](https://docs.astro.build/en/guides/i18n-routing/#browser-language-detection) in our documentation.
```
Comment (sarah11918) [inline .changeset/rude-lizards-scream.md:63]:
```suggestion
Read more about Astro's [experimental i18n routing](https://docs.astro.build/en/guides/internationalization/) in our documentation.
```
Changed the page name for docs!

File: .changeset/rude-lizards-scream.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+Enabling i18n routing also provides two new properties for browser language detection: `Astro.preferredLocale` and `Astro.preferredLocaleList`. These combine the browser's `Accept-Langauge` header, and your site's list of supported languages and can be used to automatically respect your visitor's preferred languages.
+
+```astro
+---
+const preferredLocale = Astro.preferredLocale;
+const preferredLocaleList = Astro.preferredLocaleList;
+---
+NEED A USAGE EXAMPLE HERE
```
Comment (sarah11918) [inline .changeset/rude-lizards-scream.md:60]: Can you please add a very tiny example of using these (you can choose to only use one of them in the example!) ?

Or I suggest we remove the code example entirely. Without showing usage, this doesn't add very much, and we do list the properties in the paragraph.

Either choice is fine! Whichever you prefer!

**Review by sarah11918 — COMMENTED** [review]
Last two comments from me on the changeset!

Please /ptal me if you need me to come back and look again! I'll be in the Docs PR until further notice!

**Review by sarah11918 — APPROVED** [review]
Approving the changeset!  :partying_face:

File: .changeset/rude-lizards-scream.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+Compute relative URLs for your links with `getLocaleRelativeURL` from the new `astro:i18n` module:
+
+```astro
+---
+import {getLocaleRelativeUrl} from "astro:i18n";
+const aboutUrl = getLocaleRelativeUrl("pt_Br", "about");
+---
+<p>Learn more <a href={aboutURL}>About</a> this site!</p>
+```
```
Comment (sarah11918) [inline .changeset/rude-lizards-scream.md:48]:
``````suggestion
Enable the experimental routing option by adding an `i18n` object to your Astro configuration with a default location and a list of all languages to support:

```js
// astro.config.mjs
import {defineConfig} from "astro/config";

export default defineConfig({
    experimental: {
        i18n: {
            defaultLocale: "en",
            locales: ["en", "es", "pt-br"]
        }
    }
})
```

Organize your content folders by locale depending on your `i18n.routingStrategy`, and Astro will handle generating your routes and showing your preferred URLs to your visitors.
```
├── src
│   ├── pages
│   │   ├── about.astro
│   │   ├── index.astro
│   │   ├── es
│   │   │   ├── about.astro
│   │   │   ├── index.astro
│   │   ├── pt-br
│   │   │   ├── about.astro
│   │   │   ├── index.astro
```

Compute relative URLs for your links with `getLocaleRelativeURL` from the new `astro:i18n` module:

```astro
---
import {getLocaleRelativeUrl} from "astro:i18n";
const aboutUrl = getLocaleRelativeUrl("pt-br", "about");
---
<p>Learn more <a href={aboutURL}>About</a> this site!</p>
```
``````
Updates based on last minute changes to docs PR!

File: .changeset/rude-lizards-scream.md [hunk 16]
```diff
@@ -0,0 +1,51 @@
+---
+'astro': minor
+---
+
+Experimental support for i18n routing. 
+
+Astro's experimental i18n routing allows you to add your multilingual content with support for configuring a default language and URL path convention, computing relative page URLs, and accepting preferred languages provided by your visitor's browser browser.
```
Comment (sarah11918) [inline .changeset/rude-lizards-scream.md:7]:
```suggestion
Astro’s experimental i18n routing API allows you to add your multilingual content with support for configuring a default language, computing relative page URLs, and accepting preferred languages provided by your visitor’s browser. You can also specify fallback languages on a per-language basis so that your visitors can always be directed to existing content on your site.
```
I realized we didn't mention fallback, which is a cool thing!

File: packages/astro/src/@types/astro.ts [hunk 17]
```diff
@@ -1902,6 +1989,11 @@ export type AstroFeatureMap = {
 	 * The adapter can emit static assets
 	 */
 	assets?: AstroAssetsFeature;
+
+	/**
+	 * List of features that orbit around the i18n routing
+	 */
+	i18n?: AstroInternationalisationFeature;
```
Comment (bluwy) [inline packages/astro/src/@types/astro.ts:1996]:
```suggestion
	i18n?: AstroInternationalizationFeature;
```

🇺🇸 Looks like we're also using `z` in the docs

File: packages/astro/src/@types/astro.ts [hunk 18]
```diff
@@ -1916,6 +2008,13 @@ export interface AstroAssetsFeature {
 	isSquooshCompatible?: boolean;
 }
 
+export interface AstroInternationalisationFeature {
```
Comment (bluwy) [inline packages/astro/src/@types/astro.ts:2011]:
```suggestion
export interface AstroInternationalizationFeature {
```

File: packages/astro/src/@types/astro.ts [hunk 19]
```diff
@@ -1916,6 +2008,13 @@ export interface AstroAssetsFeature {
...
+	/**
+	 * Wether the adapter is able to detect the language of the browser, usually using the `Accept-Language` header.
```
Comment (bluwy) [inline packages/astro/src/@types/astro.ts:2013]:
```suggestion
	 * Whether the adapter is able to detect the language of the browser, usually using the `Accept-Language` header.
```

File: packages/astro/src/i18n/vite-plugin-i18n.ts [hunk 20]
```diff
@@ -0,0 +1,64 @@
+import * as vite from 'vite';
+import type { AstroSettings } from '../@types/astro.js';
+
+const virtualModuleId = 'astro:i18n';
+const resolvedVirtualModuleId = '\0' + virtualModuleId;
+
+type AstroInternalization = {
```
Comment (bluwy) [inline packages/astro/src/i18n/vite-plugin-i18n.ts:7]: Needs a find+replace of `Internalization` to `Internationalization` 😅

File: packages/astro/src/core/render/core.ts
Comment (bluwy) [inline packages/astro/src/core/render/core.ts:1]: (optional) Looks like this file has many unused imports since the refactor.

File: packages/astro/src/core/endpoint/index.ts [hunk 21]
```diff
@@ -12,7 +12,8 @@ import { ASTRO_VERSION } from '../constants.js';
 import { AstroCookies, attachCookiesToResponse } from '../cookies/index.js';
 import { AstroError, AstroErrorData } from '../errors/index.js';
 import { callMiddleware } from '../middleware/callMiddleware.js';
-import type { Environment, RenderContext } from '../render/index.js';
+import { type Environment, type RenderContext } from '../render/index.js';
+import { computePreferredLocale, computePreferredLocaleList } from '../render/context.js';
```
Comment (bluwy) [inline packages/astro/src/core/endpoint/index.ts:17]: `computePreferredLocales` seems unused

Comment (ematipico) [reply packages/astro/src/core/endpoint/index.ts:17]: `computePreferredLocale` is used at line 69

Comment (bluwy) [reply packages/astro/src/core/endpoint/index.ts:17]: Ah sorry my local branch of this was outdated.

**Review by bluwy — APPROVED** [review]
